<a href="https://colab.research.google.com/github/SantiagoDemattei/dm2025a/blob/main/src/CazaTalentos/CazaTalentos_v1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Cazatalentos
## buscando a  “La verdadera mejor encestadora”

## 1.Introduccion
Este notebook analiza de forma didáctica la razón por la que se produce el overfitting.

Ver los cinco minutos de este video muy relacionado con lo que veremos
[Veritasium El Éxito ¿es Suerte o Trabajo Duro?](https://storage.googleapis.com/open-courses/austral2025-af91/labo1r/veritasium_success_paradox.html)

### 1.1 Motivacion

El siguiente desafío busca introducir el problema fundamental de la Ciencia de Datos que es el overfitting desde un ángulo distinto,  presentando los fenómenos del efecto del tamaño de la muestra,
<br> el conocido problema de las múltiples comparaciones  y “la maldición del ganador”.
<br>
<br> Aunque anecdóticametne está basado en algo deportivo, se extiende fácilmente  a :
*  la efectividad de acciones de marketing
*  el testeo de medicamentos
*  experimentos ciéntíficos https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1182327/
*  la comparación de modelos predictivos
*  la elección del mejor corte en un nodo de un árbol de decisión, etc

### 1.2 Generalidades
Vari@s cazatalent@s de la gran ciudad van recorriendo pueblos en búsqueda de las mejores jugadoras de basket para tiros libres en la Liga Femenina de Basket.
<br>El procedimiento recomendado es el siguiente : un@ cazatalentos llega a un pueblo y va a la única cancha de basket del lugar donde la están esperando las adolescentes interesadas, las hace tirar tiros libres y elige a la mejor jugadora.
<br>Posteriormente a la elección del mejor jugadora, si le parece a la entrenadora de la gran ciudad, se llevará a la jugadora a que demuestre que tan buena es.

<br>A las jugadoras los podemos equiparar con modelos predictivos.


### 2 Definición del Problema

¿Qué metodología de las que se plantearán es la que propone a la jugadora con mayor probabilidad de superar el umbral del 80% de aciertos en la nueva prueba que se hará en la gran ciudad?
<br> Cada jugadora posee un “índice de enceste”, la cazatalentos no tiene forma de conocerlos; lo único que puede hacer la cazatalentos es hacer que las jugadoras hagan tiros libres, y medir los aciertos y no aciertos.

### 2.1  Caso fallido, una jugadora se quiere pasar de lista ...

La cazatalentos llega a una pequeña localidad a la que le han dicho que hay dos adolescentes cuyas habilidades debe testear.
<br> Una es super talentosa se llama Diana Taurasi  https://en.wikipedia.org/wiki/Diana_Taurasi, nosotros sabemos que tiene un “índice de enceste” del 0.85, es decir cada vez que hace un tiro libre su probabilidad de encestar es del 85%.   Si hace 100 tiros libres, en promedio encesta 85 veces,  a veces puede ser más, a veces menos, pero en promedio encesta 85 de 100.
<br> La otra es un auténtico desastre, a quien llamaremos  Miss X, y su índice de enceste es de 0.10.  Esto quiere decir que si hace 100  tiros libres, en promedio encesta 10.
<br> La cazatalentos NO tiene forma de conocer los valores de 0.85 y 0.10, solo puede hacerlas tirar y contar aciertos.

<br>La cazatalentos llega al gimnasio,  y solo se encuentra Diana, con lo cual le indica que comience a tirar.  Al tiro 10, Diana lleva 8 encestes y 2 no-enceste.
<br>En ese momento, llega con un andar cansino Miss X y las interrumpe :  ¿cuánto va Diana ? La cazatalentos le responde  8 de 10.  Miss X solicita permiso para tirar, tira una sola vez y acierta.
<br>Ahí mismo dice : “paren las rotativas, yo soy la clara ganadora de este torneo,  mi indice de acierto es del  100% y el de Diana es del 80% ,  la matemática es clara”.
<br>La cazatalentos monta en cólera y le dice :  pues usted joven solo ha realizado un tiro, ¿Si lo hubiera errado, aceptaría decir entonces que su indice de acierto es del 0%? Ciertamente no.
<br>Pues ahora, usted deberá seguir tirando hasta completar sus 10 tiros libres, y recién ahí compararé.


### 2.2 Segundo Caso

La cazatalentos tiene el siguiente algoritmo, hace que Diana Taurasi y Miss X hagan 10 tiros libres cada una, y elige a la ganadora de esos 10 tiros libres.
¿Garantiza este método que por lo menos el 99% de las veces elige a Diana Taurasi?

<br>¿La desastrosa Miss X puede llegar a encestar los 10 tiros libres?  <br>Efectivamente sí existe esa posibilidad, pero con una muy baja probabilidad, la probabilidad es de 0.10 ^10  =  1e-10
<br>El símbolo de  ^  significa  "elevado a la potencia de"
<br>
<br>¿Cual es la probabilidad que Miss X tire una sola vez y acierte este tiro? <br>Simplemente 0.1 , una en diez, con lo cual ha tenido mucha suerte en su primer tiro, pero no la va a poder mantener !

<br>¿La adolescente Diana Taurasi puede llegar a encestar los 10 tiros libres?  Efectivamente, su probabilidad es  0.85 ^10 = 0.1968744

En el lenguaje R la instrucción  runif(10)  genera un vector con 10 números aleatorios con distribución uniforme en el intervalo [0,1], por ejemplo

In [ ]:
runif(10)

[1] 0.76354479 0.17736257 0.27908662 0.76861674 0.90664643 0.05397527
 [7] 0.71864900 0.70968798 0.92621630 0.44199519

Ahora simulamos 10000 veces la estrategia de tirar 10 tiros libres y quedarse con la ganadora.

In [ ]:
missx_ganadora <- 0

for( i  in 1:10000 ) # repite 10.000 veces el algoritmo
{
  aciertos_taurasi <- sum( runif(10) < 0.85 ) # suma la cantidad de aciertos de taurasi (cada vez que la probabilidad sea menor a 0.85 [su indice de enceste])
  aciertos_missx <- sum( runif(10) < 0.10 ) # lo mismo con miss X

  if( aciertos_taurasi < aciertos_missx )  missx_ganadora <- missx_ganadora +1 # acumulo cada vez que miss X gana
}

cat( "Miss X gano un total de : ", missx_ganadora )

Miss X gano un total de :  0

Lo que da la cantidad de 0,  o sea que en 10000 veces, siempre ganó Diana Taurasi, con lo cual la estrategia de quedarse con el mejor de 10 tiros ha funcionado, para este obvio caso donde hay apenas dos jugadoras y hay una diferencia abismal entre ellas.

En este caso NO aparece el overfitting por ningún lado

### 2.3  Aparece el Overfitting

El cazatalentos llega ahora a un pueblo donde hay 100 jugadoras los que son mucho más parejas entre sí desde el punto de vista del ratio de enceste.

Hay una jugadora, que llamaremos jugadora 1 que tiene un “índice de enceste” de  0.70.
Las 99 jugadoras restantes que llamaremos “el pelotón”  tienen los índices de enceste  { 0.501, 0.0502, 0.503, …, 0.599  }
Es decir la jugadora 1 tiene 0.70 y la mejor jugadar del pelotón tiene 0.599,  hay un poco más de 0.10 de diferencia.  La diferencia de 0.10 es significativa en basket.

¿Qué sucede con el algoritmo del cazatalentos de hacer tirar 10 tiros libres a cada una de las jugadoras y elegir a la que más encestes logró ?


In [ ]:


set.seed( 102191 )

# calcula cuantos encestes logra una jugadora con indice de enceste "prob"
# haciendo "qyt" tiros libres

ftirar  <- function( prob, qty ){
  return( sum( runif(qty) < prob ) )
}


# defino las jugadoras
jugadora1    <- 0.7
peloton    <- ( 501:599 ) / 1000
jugadoras  <- c( jugadora1, peloton )





# hago que las 100 jugadoras tiren 10 veces cada una
mapply( ftirar, jugadoras, 10 )

primera_ganadora  <- 0

for( i in 1:10000 ){  # diez mil experimentos

  vaciertos  <- mapply( ftirar, jugadoras, 10 )  # 10 tiros libres cada jugadora

  mejor  <- which.max( vaciertos )
  if( mejor == 1 )  primera_ganadora  <- primera_ganadora + 1
}

cat("\n")
cat( "En los diez mil experimentos, se detecto correctamente a la mejor jugadora la cantidad de veces:", primera_ganadora )


[1]  8  3  4  4 10  2  4  6  4  5  4  7  5  5  7  6  4  6  2  6  6  7  5  8  5
 [26]  4  3  8  6  2  5  4  6  7  6  8  9  4  8  5  4  7  6  6  7  6  7  7  3  6
 [51]  5  5  2  5  4  6  7  3  8  9  3  6  6  9  5  4  5  6  6  3  7  4  7  6  3
 [76]  6  6  7  5  2  3  6  4  7  5  7  3  6  7  7  6  7  6  9  6  7  7  6  6  9


En los diez mil experimentos, se detecto correctamente a la mejor jugadora la cantidad de veces: 1450

Aquí ya pasa algo asombroso, el overfitting en todo su esplendor.
Si hago tirar 10 tiros libres a cada uno de las 100 jugadoras, apenas 1450 veces de las 10000 este método  devuelve a la verdadera mejor, que es la jugadora con un índice de enceste de 0.70

Intuitivamente, lo que está sucediendo es que la inmensa mayoría de las veces, una de las 99 jugadoras del pelotón tiene mucha suerte y supera a la jugadora de 0.70, con lo cual se elige a una del pelotón y no a la verdadera mejor !  
<br>El tener 99 jugadoras en el pelotón hace que al ser tantas hay varias que tienen mala suerte, pero también hay varias con muy buena suerte, y la que tuvo más suerte supera a la verdadera mejor !


### 2.4 Nueva ronda

Ahora pasemos a ver aún algo más notable.
Tenemos a nuestra jugadora de 0.70, y a las 99 jugadoras del pelotón.
1.  Las hacemos tirar a cada una 10 tiros libres, y elegimos a la ganadora registrando cuántos aciertos tuvo
2.  solo a esa ganadora la hacemos tirar una nueva ronda de 10 tiros libres
3.  finalmente comparamos estos nuevos aciertos con los originales.


In [ ]:


set.seed( 102191 )

# calcula cuantos encestes logra una jugadora con indice de enceste "prob"
# haciendo "qyt" tiros libres

ftirar  <- function( prob, qty ){
  return( sum( runif(qty) < prob ) )
}


# defino los jugadoras
jugadora1 <-  0.7
peloton    <-  ( 501:599 ) / 1000
jugadoras  <-  c( jugadora1, peloton )


cat( "Aciertos Ganadora", "\t", "Nueva Ronda", "\n")

for( i in 1:10 ){ # repito el experimento 10 veces
  vaciertos  <- mapply( ftirar, jugadoras, 10 )  # cada jugadora tira 10 tiros libres
  mejor  <- which.max( vaciertos ) # tomo a la mejor jugadora, es decir, a la que encestó más
  aciertos_torneo  <- vaciertos[ mejor ] # tomo la cantidad de aciertos que tuvo esa jugadora

  aciertos_segunda  <- ftirar( jugadoras[ mejor ], 10 ) # la hago tirar otra vez 10 veces y me guardo la cantidad de veces que encestó

  cat( aciertos_torneo, "\t\t\t", aciertos_segunda, "\n" ) # imprimo ambas cantidades (primera y segunda tanda)
}

Aciertos Ganadora 	 Nueva Ronda 
10 			 5 
8 			 7 
9 			 5 
9 			 7 
9 			 5 
9 			 7 
10 			 3 
9 			 5 
10 			 7 
8 			 2 


Lo que se observa en este caso es “La maldición del ganador”,
<br>la performance que la jugadora que logró más aciertos en la competencia general luego NO LA PUEDE MANTENER,
<br>en todos los casos vemos que en la nueva ronda de diez tiene menos aciertos que en la primera ronda de diez.
<br>La primera ronda  la jugadora que resultó ganadora fue debido a la suerte, y esa suerte ya no la acompaña para la segunda ronda de 10 tiros libres.

Relacionado con las competencias Kaggle, una forma de pensar esto es  "¿Cómo le va a ir al mejor modelo del Public Leaderboard en el Private?" ,  esta es la razón por la que los puntajes que vemos en el Private son inferiores al Public.
La primer ronda puede verse como quedarnos con el mejor modelo del Public Leaderboard, la nueva ronda es como le va a ese modelo en el Private Leaderboard.
<br>Elegir el mejor modelo del Public, suele ser una pésima estrategia.

### El overfitting en su plenitud

¿Cuándo es más extremo el efecto de “La maldición del ganador” ?
<br>Supongamos por un momento que ahora, en otro pueblo,  tenemos 100 jugadoras nuevas, todas con un “índice de enceste” de 0.70 ;  recordar que ese valor jamás es conocido por la cazatalentos.

<br>Ahora nos ponemos más estrictos, y las hacemos tirar 100 veces a cada una, elegimos a la ganadora, y solo a esa ganadora la hacemos tirar 100 nuevos tiros libres.
<br> Podemos pensarlo como que la cazatalentos se lleva a la mejor jugadora a la gran ciudad, le habla maravillas de ella a la entrenadora, y la entrenadora dice “probémosla a ver si es tan buena como decís, que haga 100 tiros libres aquí y ahora”

In [ ]:


set.seed( 102191 )

# calcula cuantos encestes logra una jugadora con indice de enceste "prob"
# que hace "qyt" tiros libres

ftirar  <- function( prob, qty ){
  return( sum( runif(qty) < prob ) )
}


# defino los jugadoras
jugadoras  <- rep( 0.7, 100 ) # todas con un ratio de 0.7




cat( "Aciertos Ganadora", "\t", "Nueva Ronda", "\n")
for( i in 1:10 ){ # repito el experimento 10 veces
  vaciertos  <- mapply( ftirar, jugadoras, 100 )  # cada jugadora tira 100 tiros libres
  mejor  <- which.max( vaciertos ) # capturo aquella jugadora que hizo más aciertos en toda la tanda
  aciertos_torneo  <- vaciertos[ mejor ] # me guardo la cantidad que acertó dicha jugadora (que es la máxima en la tanda)

  aciertos_segunda  <- ftirar( jugadoras[ mejor ], 100 ) # hago tirarla otra vez 100 veces y me guardo la cantidad de aciertos

  cat( aciertos_torneo, "\t\t\t", aciertos_segunda, "\n" )
}

Aciertos Ganadora 	 Nueva Ronda 
82 			 68 
82 			 78 
83 			 78 
79 			 64 
82 			 69 
81 			 73 
78 			 71 
80 			 69 
82 			 61 
80 			 73 


Nuevamente vemos, que el puntaje alcanzado por la mejor jugadora cuando compitió contra las otras 99,  NO ES VUELTO A ALCANZAR  en la nueva ronda de 100 tiros libres que hace esa jugadora.
<br>Este efecto es exactamente el mismo que vemos cuando alguien se empecina en lograr en el Leaderboard Público el mayor puntaje de la clase,  le va muy bien en el Público, pero se derrumba catastróficamente en el Privado.
<br>Decimos que  “overfitió / sobreajustó el Leaderboard Publico”.

### 2.5 medida de la diferencia

Ahora nos preguntamos, en este caso, en promedio, cuanto más aciertos tiene la ganadora (debido a la suerte)  en la primera ronda  con respecto a la segunda ?


In [ ]:


set.seed( 102191 )

# calcula cuantos encestes logra una jugadora con indice de enceste prob
# que hace qyt tiros libres

ftirar  <- function( prob, qty ){
  return( sum( runif(qty) < prob ) )
}


# defino las jugadoras
jugadoras  <- rep( 0.7, 100 )




suma_diferencias <- 0

for( i in 1:10000 ){ # repito el experimento 10.000 veces
  vaciertos  <- mapply( ftirar, jugadoras, 100 )  # cada jugadora tira 100 tiros libres
  mejor  <- which.max( vaciertos ) # capturo aquella jugadora que hizo más aciertos en toda la tanda
  aciertos_torneo  <- vaciertos[ mejor ] # me guardo la cantidad que acertó dicha jugadora (que es la máxima en la tanda)

  aciertos_segunda  <- ftirar( jugadoras[mejor], 100 ) # hago tirarla otra vez 100 veces y me guardo la cantidad de aciertos

  suma_diferencias  <- suma_diferencias +  (aciertos_torneo - aciertos_segunda ) # acumulo la diferencia entre la primera y segunda tanda
}


cat( "Diferencia promedio:",  suma_diferencias / 10000 ) # calculo el promedio


Diferencia promedio: 11.0609

Exactamente este mismo efecto se da en el marketing digital.  Si en un A/B testing se prueban cien alternativas y se elige a la que tiene más efectividad, se observará que en un nuevo experimento disminuirá el rendimiento de esa alternativa.

También sucede si se prueban muchos medicamentos al mismo tiempo,   el que resulte más efectivo disminuirá su efectividad en la nueva prueba.

Este problema es llamado “El problema de las múltiples comparaciones”

### 2.6 una sola jugadora MUY buena

Sexto caso,  ¿cuándo se atenúa/desaparece  el overfitting?
Supongamos 99 jugadoras con un índice de enceste de 0.60   y  ahora sumamos a la adolescente Diana Taurasi con su superlativo índice de enceste del 0.85.
Las hacemos tirar a todas 100 tiros libres, llevamos a la ganadora a la gran ciudad, y allí le contamos a la entrenadora de nuestra ganadora Diana Taurasi.

In [ ]:


set.seed( 102191 )

# calcula cuantos encestes logra una jugadora con indice de enceste prob
# que hace qyt tiros libres

ftirar  <- function( prob, qty ){
  return( sum( runif(qty) < prob ) )
}


# defino las jugadoras
taurasi    <- 0.85
peloton   <- rep( 0.6, 99 )  # jugadoras identicas
jugadoras <- c( taurasi, peloton )



cat( "Aciertos Ganadora", "\t", "Nueva Ronda", "\n")

for( i in 1:10 ){ # repito 10 veces
  vaciertos  <- mapply( ftirar, jugadoras, 100 )  # cada jugadora tira 100 tiros libres
  mejor  <- which.max( vaciertos ) # capturo aquella jugadora que hizo más aciertos en toda la tanda
  aciertos_torneo  <- vaciertos[ mejor ] # me guardo la cantidad que acertó dicha jugadora (que es la máxima en la tanda)

  aciertos_segunda  <- ftirar( jugadoras[ mejor ], 100 ) # hago tirarla otra vez 100 veces y me guardo la cantidad de aciertos

  cat(aciertos_torneo, "\t\t\t", aciertos_segunda, "\n" )
}



Aciertos Ganadora 	 Nueva Ronda 
88 			 91 
85 			 86 
79 			 91 
86 			 81 
87 			 87 
84 			 84 
86 			 83 
87 			 86 
85 			 77 
92 			 83 


En este caso observamos dos cosas, en primer lugar SIEMPRE la ganadora del torneo fue Diana Taurasi, y en segundo  cuando tuvo que mostrar en la ciudad su performance, fue la misma que en el primer torneo.  
<br>Es tan buena Taurasi que le ganó a la que tuvo más suerte del pelotón, el efecto suerte no está afectando a Taurasi porque está sola, y gana por su superioridad, no por la suerte.

### 2.7 Efectos del tamaño de la ronda

La cazatalentos llega ahora a un pueblo donde hay 100 jugadoras las que son mucho más parejas entre sí desde el punto de vista del ratio de enceste.
Hay un jugadora, que llamaremos jugadora 1 que tiene un “índice de enceste” de  0.70 ,  es una jugadora buena pero dista mucho de ser Diana Taurasi, jamás estará en el "WNBA Hall of Fame".

Las 99 jugadoras restantes que llamaremos “el pelotón”  tienen los índices de enceste  { 0.501, 0.0502, 0.503, …, 0.599  }  Es decir la jugadora 1 tiene 0.70 y la mejor jugadora del pelotón tiene 0.599,  o sea hay un poco más de 0.10 de diferencia.  La diferencia de 0.10 es significativa en basket.

¿Cuántos tiros libres debe la cazatalentos pedirles que haga cada jugadora para que  si elige a la mejor jugadora tiene la certeza de llevarse a “la verdadero mejor” el 99% de los casos ?


In [ ]:
# intencionalmente la mejor jugadora va al final de la lista de jugadoras
# porque la funcion which.max() de R hace trampa
# si hay un empate ( dos máximos) se queda con la que esta primera en el vector


set.seed( 102191 )

# calcula cuantos encestes logra una jugadora con indice de enceste prob
# que hace qyt tiros libres

ftirar  <- function( prob, qty ){
  return( sum( runif(qty) < prob ) )
}


# defino los jugadoras
taurasi    <- 0.7
peloton    <- ( 501:599 ) / 1000
jugadoras  <- c( peloton, taurasi) # intencionalmente la mejor esta al final


for( tiros_libres in c(10, 20, 50, 100, 200, 300, 400, 415, 500, 600, 700, 1000 ) ){ # cantidades de tiros a hacer por cada jugadora

  ultima_ganadora <- 0

  for( i in 1:10000 ) { # diez mil experimentos

    vaciertos  <- mapply( ftirar, jugadoras, tiros_libres ) # hago tirar a cada jugadora la cantidad "tiros_libres"
    mejor  <- which.max( vaciertos ) # me guardo aquella jugadora que obtuvo la mayor cantidad de aciertos (el índice de la lista "jugadoras")

    # taurasi (= ultima_jugadora) está en la posición 100 de la lista "jugadoras"
    if( mejor == 100 )  ultima_ganadora  <- ultima_ganadora + 1 # acumulo cada vez que taurasi gana
  }

  cat( tiros_libres, "\t", ultima_ganadora/10000, "\n" ) # imprimo cantidad de tiros libres y el promedio de veces que ganó taurasi
}



10 	 0.0299 
20 	 0.0868 
50 	 0.268 
100 	 0.5448 
200 	 0.8532 
300 	 0.9579 
400 	 0.9877 
415 	 0.9911 
500 	 0.9964 
600 	 0.9995 
700 	 0.9999 
1000 	 1 


Haciendo tirar 415 tiros libres a cada una de los 100 jugadoras,
<br> y quedándose con la ganadora de ese torneo,
<br> en el 99.10%   ( 0.9910 )  de los casos se está segura que esa es "la verdadera mejor".


## 3 Conclusiones

El overfitting aparece cuando se comparan muchas jugadoras (modelos predictivos)  muy parecidas entre sí.  
<br> En el modelado predictivo es la situación más común que se tengan modelos con similar poder predictivo ya que generalmente solo difieren apenas en valores de los hiperparámetros o algunas columnas nuevas con feature engineering.

Las jugadoras no son determinísticas, sino probabilísticas, con lo cual al comparar muchas parecidas la ganadora es quien además de su talento tuvo más suerte en ese torneo.  La jugadora ganadora ganó por mera suerte, y  NO puede sostener esa suerte en un nuevo torneo, por lo que el puntaje del primer torneo es mentiroso, no se sostiene en el nuevo torneo.

El overfitting se atenúa o puede llegar a desaparecer del todo, si en el torneo hay una jugadora ampliamente superior al resto. Esa jugadora SI mantiene su performance de un torneo al otro, ya que su puntaje en el primer torneo fue alto porque es realmente buena, porque no es el máximo de muchos parecidas a ella.

La forma de combatir el overfitting es hacer torneos de muchos tiros libres.
<br>Para estimar que tan buena es la jugadora ganadora de un torneo, hay que hacerla tirar nuevamente tiros libres.
<br>El secreto está en hacer más de un torneo.


## 4 Desafio Cazatalentos Ordenamiento

### 4.1 Enunciado

Se reúne la entrenadora  en la gran ciudad con algunas cazatalentos que visitaron regiones distintas.

Entrenadora:  Bienvenidas a todas, veo que cada una de ustedes ha traído la estadística de su candidata. Sin embargo, yo debo ordenarlas de mejor a peor, y no me alcanza con que dos de ustedes me digan que que sus candidatas encestaron 80/100, necesito que me relaten exactamente el procedimiento que han seguido para elegir a la mejor jugadora.

Cazatalentos 1:  Discúlpeme entrenadora, si dos de nosotras le presentamos casos que en distintos pueblos han encestado exactamente 80 de 100, probabilisticamente esas adolescentes son equivalentes. Es cierto que luego aquí en la gran ciudad el dia de la prueba una podrá encestar 77 de 100, la otra 83,  pero eso es la parte natural del random, ya sabemos que la distribución es binomial.  Ambas son equivalentes, no importa que pasó con el resto de las jugadoras de su pueblo.

Entrenadora: Estimada “Cazatalentos 1”  tú estas equivocada, no consideras El Problema de las Multiples Comparaciones; elegir cuales de esas adolescentes vendrán a la gran ciudad a ser evaluadas es mi decisión. Comienza tú contando detalladamente el proceso que has seguido.
4



**Cazatalentos 1**: Llegué a la localidad habían 100 adolescentes en el gimnasio, hice tirar a cada una 100 tiros, solo una obtuvo 80 aciertos, luego un par 79, otro par 78, y así descendían. La que logró 80 aciertos es mi candidata.


**Cazatalentos 2:** Llegué a una pequeña ciudad y habían 200 adolescentes en el gimnasio, hice tirar a cada una 100 tiros, solo una obtuvo 80 aciertos, seis 79,  cinco 78, y así descendían. La que logró 80 aciertos es mi candidata.

**Cazatalentos 3:** Yo poseo 25 años de experiencia como cazatalentos y tengo muy buena intuición  para detectar el talento. Llegué a un pueblo donde habían 100 adolescentes, me senté en las gradas del gimnasio y las observé detenidamente mientras precalentaban sin tirar al aro.  A los 10 minutos, me paro, y señalo a una de ellas y le digo :  'tú, la número 13, a tirar 100 tiros; el resto gracias por haber venido pero se pueden ir cada una a su casa”. Y ahí al tirar esos 100 tiros logró 80 aciertos y esa es mi candidata.

**Cazatalentos 4:** Llegué a minúsculo pueblo con un aro reglamentario en la pared de una escuela, habían solamente 2 adolescentes, hice tirar a cada una 100 tiros, una obtuvo 80 y la otra 75.

**Cazatalentos 5:** Le dice a la entrenadora : “fui a una localidad de 500 habitantes, había una sola adolescente interesada, la hice tirar 100 tiros libres, y encestó 80,  tengo una filmación con mi celular de los 100 tiros, sin edición ni cortes”
En realidad, esa adolescente es la sobrina de la cazatalentos, la hizo tirar 10 rondas de 100 tiros cada una, grabó todas, pero tramposamente solo trae la grabación de la mejor ronda a la entrenadora.
Los resultados obtenidos por su sobrina son


| ronda | Tiros | Encestes |
| ---: | ---: | ---: |
| 1 | 100 | 68 |
| 2 | 100 | 74 |
| 3 | 100 | 78 |
| 4 | 100 | 70 |
| 5 | 100 | 68 |
| 6 | 100 | 63 |
| 7 | 100 | 80 |
| 8 | 100 | 68 |
| 9 | 100 | 67 |
| 10 | 100 | 65 |


(esta deshonestidad de la Cazatalentos 5 no debería parecerle extraña ni causarle gracia, es exactamente lo mismo que elegir como submit final el que le fue mejor en el Public Leaderboard,  algo muy común entre l@s alumn@s ...)
<br>Por una persona espía en el pueblo, la entrenadora está al tanto de lo ocurrido y conoce la tabla anterior, usará dicha información para tomar la decisión, aunque no le dirá nada, por esta vez, a la cazatalentos.

### 4.2 Desafio

Podemos suponer que la supertalentosa Diana Taurasi con su “índice de enceste” del 0.85 es superior a todas las jugadoras evaluadas en los pueblos por las cazatalentos.

Desafío : ¿Si usted fuera la entrenadora, como ordena el valor esperado de encestes en una ronda de 100 tiros libres  a cada una de las candidatas propuestas por las cinco cazatalentos ?


donde   Ci  < Cj   sii   prob(  #Encestes Ci  <  #Encestes Cj    | 100 tiros cada uno )  >  0.5
es decir que, haciendo 100 tiros libres,  más de la mitad de las veces,  Ci va a lograr menos encestes que Cj

La entrenadora de la gran ciudad NO quiere traer a la elegida de cada cazatalentos y hacerlas tirar 100 tiros delante de ella para constatar cual logra más encestes en esa ronda. Ella quiere, con la información que le brindó cada cazatalentos sobre el procedimiento y los resultados, elegir la mejor encestadora.


In [ ]:
bado

ERROR: Error: object 'bado' not found


### 4.3  Reglas del Desafio


*  La fecha de finalización del desafío es el sabado 02 de agosto a las 19:00
Usted podrá trabajar en forma individual o formando grupo de dos personas.
*  Entrega
      * la entrega se realiza con un solo mensaje en Zulip canal #CazaTalentos
         * Ordenamiento,  por ejemplo     C3 <  C5 <  C1 < C4 < C2
         * Explicacion
      * Si necesitará apoyarse en scripts para alguna de las comparaciones
          * link a carpeta en GitHub
      * Podrá enviar varios mensajes, el único que tendrá valor será el último

*  Recompensa:
     * Todos los integrantes de todos los grupos que logren un ordenamiento perfecto sumarán a su nota final el valor de 1.0 ,  siempre la nota máxima posible de toda la asignatura será de 10.0
     * Un ordenamiento que no sea el correcto, suma 0 puntos.
*  Será muy valioso el intercambio de ideas en Zulip channel  #CazaTalentos  topic  Intercambio de Ideas, este desafío suele ser muy polémico.
*  La participación en este desafío es opcional.


# **Resolución**

In [20]:
set.seed(102191)

# función para tirar
ftirar <- function(prob, qty) {
  return(sum(runif(qty) < prob))
}

## **Análisis de cada Cazatalentos**
Para resolver este problema, necesito estimar cuál es la probabilidad REAL
de enceste de cada candidata, no solo confiar en el 80/100 que todas lograron

El truco está en entender que:
- Si una jugadora gana entre muchas, probablemente tuvo suerte
- Si una jugadora es elegida sin competir, su resultado es más confiable
- Si reportan solo la mejor de varias rondas, es trampa

## **Cazatalentos 5**

In [21]:
# C5 es fácil de analizar porque conozco TODAS sus rondas
aciertos_c5 <- c(68, 74, 78, 70, 68, 63, 80, 68, 67, 65)
promedio_c5 <- sum(aciertos_c5) / 10
cat("C5 - Promedio real de todas las rondas:", promedio_c5, "\n")
cat("C5 - Probabilidad real estimada:", promedio_c5/100, "\n")

C5 - Promedio real de todas las rondas: 70.1 
C5 - Probabilidad real estimada: 0.701 


## **Simulo competencias para estimar las probabilidades reales**

Para los otros casos, voy a simular qué probabilidad real debería tener una jugadora para que, compitiendo contra N otras, tenga buenas chances de ganar con 80/100

In [22]:
# función para simular un torneo y ver si una jugadora con prob_real gana con 80+ aciertos
simular_torneo <- function(prob_real, n_competidoras, tiros_por_jugadora) {

  aciertos_candidata <- ftirar(prob_real, tiros_por_jugadora) # la jugadora que me interesa

  # las otras jugadoras del pelotón (asumo que tienen probabilidades menores)
  max_aciertos_peloton <- 0
  for(i in 1:n_competidoras) {
    prob_peloton <- runif(1, 0.5, prob_real - 0.05)  # hago que sean un poco peores que la candidata
    aciertos_peloton <- ftirar(prob_peloton, tiros_por_jugadora)
    if(aciertos_peloton > max_aciertos_peloton) {
      max_aciertos_peloton <- aciertos_peloton
    }
  }

  # devuelvo TRUE si mi candidata ganó con 80+ aciertos
  return(aciertos_candidata >= 80 && aciertos_candidata > max_aciertos_peloton)
}

In [23]:
# Para cada cazatalentos, encuentro qué probabilidad real hace que sea creíble ganar con 80/100

# C1: 100 competidoras
exitos_c1 <- 0
for(prob_test in seq(0.75, 0.85, 0.01)) {
  exitos <- 0
  for(i in 1:1000) {
    if(simular_torneo(prob_test, 99, 100)) exitos <- exitos + 1
  }
  if(exitos > 100) {  # si es creíble que gane con esta probabilidad
    prob_real_c1 <- prob_test
    break
  }
}

# C2: 200 competidoras (más difícil ganar sin suerte)
for(prob_test in seq(0.78, 0.88, 0.01)) {
  exitos <- 0
  for(i in 1:1000) {
    if(simular_torneo(prob_test, 199, 100)) exitos <- exitos + 1
  }
  if(exitos > 100) {
    prob_real_c2 <- prob_test
    break
  }
}

# C3: sin competencia, resultado directo
prob_real_c3 <- 80/100  # es lo que realmente tiró

# C4: Solo 2 competidoras
for(prob_test in seq(0.72, 0.82, 0.01)) {
  exitos <- 0
  for(i in 1:1000) {
    if(simular_torneo(prob_test, 1, 100)) exitos <- exitos + 1
  }
  if(exitos > 200) {  # más fácil con menos competencia
    prob_real_c4 <- prob_test
    break
  }
}

# C5: ya la calculé antes
prob_real_c5 <- promedio_c5/100

cat("Probabilidades reales estimadas:\n")
cat("C1 (100 competidoras):", prob_real_c1, "\n")
cat("C2 (200 competidoras):", prob_real_c2, "\n")
cat("C3 (sin competencia):", prob_real_c3, "\n")
cat("C4 (2 competidoras):", prob_real_c4, "\n")
cat("C5 (trampa):", prob_real_c5, "\n")

Probabilidades reales estimadas:
C1 (100 competidoras): 0.75 
C2 (200 competidoras): 0.78 
C3 (sin competencia): 0.8 
C4 (2 competidoras): 0.77 
C5 (trampa): 0.701 


## **Simulo rendimiento futuro**
Ahora simulo como les iría en una nueva ronda de 100 tiros

In [24]:
suma_c1 <- 0
suma_c2 <- 0
suma_c3 <- 0
suma_c4 <- 0
suma_c5 <- 0

for(i in 1:10000) {
  suma_c1 <- suma_c1 + ftirar(prob_real_c1, 100)
  suma_c2 <- suma_c2 + ftirar(prob_real_c2, 100)
  suma_c3 <- suma_c3 + ftirar(prob_real_c3, 100)
  suma_c4 <- suma_c4 + ftirar(prob_real_c4, 100)
  suma_c5 <- suma_c5 + ftirar(prob_real_c5, 100)
}

promedio_futuro_c1 <- suma_c1 / 10000
promedio_futuro_c2 <- suma_c2 / 10000
promedio_futuro_c3 <- suma_c3 / 10000
promedio_futuro_c4 <- suma_c4 / 10000
promedio_futuro_c5 <- suma_c5 / 10000

cat("Aciertos esperados en nueva ronda de 100 tiros:\n")
cat("C1:", promedio_futuro_c1, "\n")
cat("C2:", promedio_futuro_c2, "\n")
cat("C3:", promedio_futuro_c3, "\n")
cat("C4:", promedio_futuro_c4, "\n")
cat("C5:", promedio_futuro_c5, "\n")

Aciertos esperados en nueva ronda de 100 tiros:
C1: 75.0133 
C2: 77.9905 
C3: 79.9245 
C4: 76.9999 
C5: 70.1173 


Ahora comparo de a pares para determinar el orden de las candidatas. Esta simulación consiste en elegirlas de a pares y hacer un torneo de 100 tiros entre ellas. Repito dicho torneo 10.000 en cada par y me quedo con la mejor. De esta forma, las puedo ordenar de menor a mayor.

In [25]:
matriz_victorias <- matrix(0, 5, 5)
probs_reales <- c(prob_real_c1, prob_real_c2, prob_real_c3, prob_real_c4, prob_real_c5)

# hago todas las comparaciones de a pares posibles
for(i in 1:5) {
  for(j in 1:5) {
    if(i != j) {
      victorias_i <- 0
      for(sim in 1:10000) {
        aciertos_i <- ftirar(probs_reales[i], 100)
        aciertos_j <- ftirar(probs_reales[j], 100)
        if(aciertos_i > aciertos_j) victorias_i <- victorias_i + 1
      }
      matriz_victorias[i,j] <- victorias_i # me guardo la cantidad de veces que le ganó i a j
      cat("C", i, " vs C", j, ": C", i, " gana ", victorias_i, " veces de 10000\n", sep="")
    }
  }
}

C1 vs C2: C1 gana 2810 veces de 10000
C1 vs C3: C1 gana 1781 veces de 10000
C1 vs C4: C1 gana 3389 veces de 10000
C1 vs C5: C1 gana 7555 veces de 10000
C2 vs C1: C2 gana 6668 veces de 10000
C2 vs C3: C2 gana 3229 veces de 10000
C2 vs C4: C2 gana 5346 veces de 10000
C2 vs C5: C2 gana 8815 veces de 10000
C3 vs C1: C3 gana 7732 veces de 10000
C3 vs C2: C3 gana 5980 veces de 10000
C3 vs C4: C3 gana 6661 veces de 10000
C3 vs C5: C3 gana 9398 veces de 10000
C4 vs C1: C4 gana 6011 veces de 10000
C4 vs C2: C4 gana 4043 veces de 10000
C4 vs C3: C4 gana 2675 veces de 10000
C4 vs C5: C4 gana 8422 veces de 10000
C5 vs C1: C5 gana 1964 veces de 10000
C5 vs C2: C5 gana 847 veces de 10000
C5 vs C3: C5 gana 481 veces de 10000
C5 vs C4: C5 gana 1208 veces de 10000


## **Conclusión**

Luego de compararlas de a pares, determino el orden segun esos resultados.

In [26]:
puntos <- c(0, 0, 0, 0, 0)
for(i in 1:5) {
  for(j in 1:5) {
    if(i != j && matriz_victorias[i,j] > 5000) {
      puntos[i] <- puntos[i] + 1
    }
  }
}

cat("\nPuntuación total (a cuántas candidatas le gana cada una):\n")
for(i in 1:5) {
  cat("C", i, ": ", puntos[i], " victorias\n", sep="")
}

# Ordenamiento basado en puntuación (de menor a mayor puntos)
cat("\nOrdenamiento final de PEOR a MEJOR (basado en comparaciones pareadas):\n")

# Encontramos manualmente el orden
candidatas_ordenadas <- character(5)
puntos_usados <- rep(FALSE, 5)

for(posicion in 1:5) {
  min_puntos <- 5  # Máximo posible + 1
  candidata_min <- 0

  for(i in 1:5) {
    if(!puntos_usados[i] && puntos[i] < min_puntos) {
      min_puntos <- puntos[i]
      candidata_min <- i
    }
  }

  candidatas_ordenadas[posicion] <- paste0("C", candidata_min)
  puntos_usados[candidata_min] <- TRUE
}

cat(paste(candidatas_ordenadas, collapse=" < "), "\n")


Puntuación total (a cuántas candidatas le gana cada una):
C1: 1 victorias
C2: 3 victorias
C3: 4 victorias
C4: 2 victorias
C5: 0 victorias

Ordenamiento final de PEOR a MEJOR (basado en comparaciones pareadas):
C5 < C1 < C4 < C2 < C3 


Explicación:
- C5 es la peor porque hizo trampa (eligió la mejor de 10 rondas)
- C2 es peor que C1 porque compitió contra más jugadoras (200 vs 100)
- C1 sufrió la 'Maldición del ganador' por competir contra 99
- C4 compitió solo contra 1, menos efecto suerte
- C3 es la mejor porque no hubo competencia, resultado puro